File: src/main/java/com/calincosma/mavenizer/service/NexusService.java
package com.calincosma.mavenizer.service;

import com.calincosma.mavenizer.domain.Jar;
import com.calincosma.mavenizer.domain.nexus.Artifact;
import com.calincosma.mavenizer.domain.nexus.NexusResponse;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.sun.jersey.api.client.Client;
import com.sun.jersey.api.client.ClientResponse;
import com.sun.jersey.api.client.WebResource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.xml.bind.annotation.adapters.HexBinaryAdapter;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.MessageFormat;
import java.util.List;

public class NexusService {
	
	public static final String NEXUS_URL_SEARCH_SHA1 = "http://search.maven.org/solrsearch/select?q=1:\"{0}\"&rows=50&wt=json";
	public static final String NEXUS_URL_SEARCH_CLASS = "http://search.maven.org/solrsearch/select?q=fc:\"{0}\"&rows=50&wt=json";
	
	public static final Logger LOGGER = LoggerFactory.getLogger(NexusService.class);
	
	public String calcSHA1(File file) throws IOException, NoSuchAlgorithmException {
		MessageDigest sha1 = MessageDigest.getInstance("SHA-1");
		try (InputStream input = new FileInputStream(file)) {
			
			byte[] buffer = new byte[8192];
			int len = input.read(buffer);
			
			while (len != -1) {
				sha1.update(buffer, 0, len);
				len = input.read(buffer);
			}
			
			return new HexBinaryAdapter().marshal(sha1.digest());
		}
	}
	
	
	
	public String nexusSearch(String url) {
		Client client = Client.create();
		WebResource webResource = client.resource(url);
		ClientResponse response = webResource.accept("application/json").get(ClientResponse.class);
		if (response.getStatus() != 200) {
			throw new RuntimeException("Failed : HTTP error code : " + response.getStatus());
		}
		return response.getEntity(String.class);
	}
	
	
	
	public Artifact findMavenArtifact(Jar jar) {
		LOGGER.debug("Searching Maven Central by SHA1 for " + jar);
		try {
			String sha1 = calcSHA1(new File(jar.getFullPath()));
			String url = MessageFormat.format(NEXUS_URL_SEARCH_SHA1, sha1);
			LOGGER.debug("Maven Central URL: " + url);
			String json = nexusSearch(url);
			Gson gson = new GsonBuilder().create();
			NexusResponse nexusResponse = gson.fromJson(json, NexusResponse.class);
			
			if (nexusResponse.getResponseHeader().getStatus() != 0)
				throw new RuntimeException("SHA1 checksum search failed with status: " + nexusResponse.getResponseHeader().getStatus());
			
			if (nexusResponse.getResponse().getNumFound() > 1)
				throw new RuntimeException("SHA1 checksum search returned too many results");
			
			if (nexusResponse.getResponse().getNumFound() == 0) {
				LOGGER.info("No Maven artifact found for " + jar.getFullPath());
				return null;
			}
			
			LOGGER.info("Found Maven artifact for " + jar.getFullPath() + "; " + nexusResponse.getResponse().getArtifacts().get(0));
			return nexusResponse.getResponse().getArtifacts().get(0);
		} catch (Exception e) {
			LOGGER.error("Error while searching Maven artifact for " + jar.getFullPath(), e);
			return null;
		}
	}
	
	
	
	public List<Artifact> findByClass(String className) {
		LOGGER.debug("Searching Maven Central by class name for " + className);
		try {
			String url = MessageFormat.format(NEXUS_URL_SEARCH_CLASS, className);
			LOGGER.debug("Maven Central URL: " + url);
			String json = nexusSearch(url);
			Gson gson = new GsonBuilder().create();
			NexusResponse nexusResponse = gson.fromJson(json, NexusResponse.class);
			
			if (nexusResponse.getResponseHeader().getStatus() != 0)
				throw new RuntimeException("Class name search failed with status: " + nexusResponse.getResponseHeader().getStatus());
			
			LOGGER.info("Found Maven " + nexusResponse.getResponse().numFound + " artifacts for " + className);
			return nexusResponse.getResponse().getArtifacts();
		} catch (Exception e) {
			LOGGER.error("Error while searching Maven artifact for " + className, e);
			return null;
		}
	}
}

File: src/main/kotlin/com/calincosma/mavenizer/service/NexusService.kt
package com.calincosma.mavenizer.service

import com.calincosma.mavenizer.model.NexusArtifact
import com.calincosma.mavenizer.model.NexusResponse
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json
import org.slf4j.LoggerFactory

class NexusService {

	private val NEXUS_URL_SEARCH = "https://search.maven.org/solrsearch/select"

	val LOGGER = LoggerFactory.getLogger(NexusService::class.java)




	private suspend fun nexusSearchBySha1(sha1: String) : NexusResponse = nexusSearch("1", sha1)

	private suspend fun nexusSearchByClassName(className: String) : NexusResponse = nexusSearch("fc", className)

	private suspend fun nexusSearch(paramPrefix: String, param: String): NexusResponse {

		val client = HttpClient(CIO) {
			install(ContentNegotiation) {
				json(Json {
					prettyPrint = true
					isLenient = true
					ignoreUnknownKeys = true
				})
			}
			install(Logging) {
				logger = Logger.DEFAULT
				level = LogLevel.ALL
			}
		}
		val response: HttpResponse = client.get {
			url(NEXUS_URL_SEARCH)
			parameter("q", "$paramPrefix:$param")
			parameter("rows", "50")
			parameter("wt", "json")
		}
		if (response.status.isSuccess())
			return response.body()


		LOGGER.error(response.bodyAsText())
		throw RuntimeException("Nexus search failed. HTTP error code: ${response.status}")
	}


	suspend fun findNexusArtifacts(sha1: String): List<NexusArtifact> {
		LOGGER.debug("Searching Maven Central by SHA1 for $sha1")
		try {
			val nexusResponse: NexusResponse = nexusSearchBySha1(sha1)
			if (nexusResponse.responseHeader.status != 0) throw RuntimeException("SHA1 checksum search failed with status: ${nexusResponse.responseHeader.status}")
			if (nexusResponse.response.numFound > 0) {
				LOGGER.debug("Found Maven artifact ${nexusResponse.response.artifacts.get(0)} for $sha1")
				return nexusResponse.response.artifacts
			}
			LOGGER.warn("No Maven artifact found for $sha1")
		} catch (e: Exception) {
			LOGGER.error("Error while searching Maven artifact for $sha1", e)
		}

		return listOf()
	}


	suspend fun findByClass(className: String): List<NexusArtifact?>? {
		LOGGER.debug("Searching Maven Central by class name for $className")
		return try {
			val nexusResponse: NexusResponse = nexusSearchByClassName(className)
			if (nexusResponse.responseHeader.status != 0) throw RuntimeException("Class name search failed with status: " + nexusResponse.responseHeader.status)
			LOGGER.info("Found Maven " + nexusResponse.response.numFound + " artifacts for " + className)
			nexusResponse.response.artifacts
		} catch (e: java.lang.Exception) {
			LOGGER.error("Error while searching Maven artifact for $className", e)
			null
		}
	}
}